### DISCUSSION TITLE
Discount Vouchers in Synchronous Tax Webhook

### DISCUSSION THREAD

**Discussion by rrrnld** [saleor/saleor#12607] [Q&A]
I'm using the latest Saleor 3.11 and have an external app that calculates taxes. Up until now I simply ignored discount vouchers. This was tested with vouchers for individual products, which worked fine, because I basically calculate taxes line-by-line, setting `totalNetAmount`, `totalGrossAmount`, `taxRate` accordingly.

Lets take the request shown in the docs as an example: https://docs.saleor.io/docs/3.x/developer/extending/apps/synchronous-webhooks/tax-webhooks#request-format

I go through each line individually, calculating taxes for different product types:

```
if checkout.IncludedTaxesInPrices {
	totalGrossAmount = lineTotal
	totalNetAmount = lineTotal.Div(taxRate)
} else {
	totalGrossAmount = lineTotal.Mul(taxRate)
	totalNetAmount = lineTotal
}
// append to lines
```

Then I calculate  the shipping tax rate and return everything like so:

```
return CheckoutTaxData{
	Lines: lines,

	ShippingTaxRate:          shippingTaxRate,
	ShippingPriceGrossAmount: shippingPriceGrossAmount,
	ShippingPriceNetAmount:   shippingPriceNetAmount,
}
```

I do not calculate any total and would expect the "new user discount" to be subtracted from `checkout.totalPrice.amount` without any further step, because the webhook does not allow me to manipulate the total.

However, I noticed that the discount is not taken into consideration. I have a percentage-based discount that is calculated correctly (which I can check after the webhook has run in `{ checkout { discount } }`), but `{ checkout { totalPrice } }` is not reduced accordingly.

I know the documentation mentions different kinds of vouchers, but it does not mention vouchers valid for all product types: https://docs.saleor.io/docs/3.x/developer/extending/apps/synchronous-webhooks/tax-webhooks#checkout-taxes

Do I need to take care to apply the voucher myself? Where and when would I apply it? How can I know which kind of voucher it is?

Thanks in advance.

**Comment by rrrnld** [comment]
I just saw this line in the docs:

> List of discounts assigned to checkout. (Those discounts should be applied by a tax App.)

So I just subtract it from the individual lines, spreading it evenly? That feels a bit odd, because the calculation in the other direction (going from a percentage based voucher to an absolute amount) has already been done by saleor…

**Comment by rrrnld** [comment]
I'd like to clarify my problem in the hopes that this is useful feedback to the saleor team:

We have a range of products with complex tax rules for VAT. Some taxes are calculated based on an order's destination country, some based on the source country. Depending on the destination country, some taxes are applied by us at all. This is why we chose to implement our taxes first as a plugin, and now as an app, as per your suggestion in another thread.

The webhook response format for is very simple (taken from https://docs.saleor.io/docs/3.x/developer/extending/apps/synchronous-webhooks/tax-webhooks#response-format):

``` json
{
 "shipping_tax_rate": "10"
 "shipping_price_gross_amount": 59.17,
 "shipping_price_net_amount": 53.79,
  "lines": [
    {
      "tax_rate": "10",
      "total_gross_amount": 17.49,
      "total_net_amount": 15.91
    },
    {
      "tax_rate": "10",
      "total_gross_amount": 55.40,
      "total_net_amount": 50.36
    }
  ],
}
```

However, this conflates discounts and taxes. This gets problematic because currently (we're using Saleor 3.11) discounts for individual products are handled differently to taxes on an entire order. As described above and as mentioned in the docs, discounts for the entire order need to be applied by the tax app. We chose the strategy of evenly spreading the amount over the different lines, but this does not set `unitPrice` / `undiscountedUnitPrice` correctly.

Being able to set `taxRate`, `undiscountedUnitPrice`, `unitPrice` and `totalPrice` in the response could be a possible solution. It would also be desirable to unify the handling of the different discount types.

In any case, thanks for providing Saleor as an open source product, it is much appreciated.

**Comment by korycins** [comment]
Hey @heyarne,
Thank you for the feedback ❤️  I hope to provide some clarification.

The difference in how we include discounts in prices is related to the information we have at the time of sending the webhook. If we know that a voucher is dedicated to a specific product, we subtract it from that specific line. The same goes for shipping. If we have a voucher for shipping, we know how to subtract it. We do not propagate the voucher for the entire order, as different tax providers handle this differently.

For example, for TaxJar, you would probably subtract the entire discount value from each line. However, for Avatax, you would add the field `discount: <discount_amount>` and `discounted: true` to each line. In this case, Avatax would handle the propagation of the discount across lines (and shipping).

There is a bug on our side. We should use the data you provided to calculate the `unitPrice`. The `undiscountedX` prices should also be calculated using the `tax_rate` returned in the response.

Thank you again for raising this issue. I will create a task to address it.

**Reply by korycins** [reply]
https://github.com/saleor/saleor/issues/12691

**Reply by rrrnld** [reply]
@korycins  Thank you for your response! I'm not using any upstream tax provider. I do the tax calculation within a custom app and communicate only between saleor and the app.

If I understand you correctly, you introduce this special case for some discounts because some third parties need special treatment. However the current implementation does not designate it as special treatment, but instead makes this the normal case. I feel like explicitly returning `undistountedUnitPrice`, `unitPrice`, `totalPrice` could be a way to provide a unified interface for all cases, but I'll gladly admit that I'm not familiar with any of the upstream tax providers you mentioned. Maybe there is still time to iron this out before the tax API is finalized.

With regards to your suggestion of re-caclulating the `undiscountedX` prices by using the tax rate: If I understand the suggestions correctly this could lead to rounding errors, causing fluctuating unit prices on different invoices. Is this a valid concern?

**Reply by korycins** [reply]
@heyarne Thank you for your feedback.

We aimed to create a generic implementation that would be suitable for all use cases related to tax calculation. This is why we treat entire voucher discounts differently.

Regarding the possible returning `undiscountedUnitPrice` and `unitPrice` in the App response, we wanted to keep the return response as simple as possible. All of these values can be calculated by Saleor based on the current response's data. However, I will pass your suggestions on to the team.

Regarding `undiscountedX` and the tax rate, I believe that the invoice should only contain the prices after all discounts have been applied. Details about prices before discounts are not relevant since they are not the amounts paid by the user.

Based on the provided tax rate, we should calculate the net and gross amounts of `undiscountedTotalX` and attach them to the line. The tax rate should also be used to calculate the undiscountedUnitPrice. You should not have any problems with rounding between unit and total.

When your App returns the total with taxes, the total price should be used to calculate the line price by Saleor. If the returned `total` divided by `line.quantity` gives us an amount with more decimal places than the currency, then `unitPrice` will be rounded.

**Reply by patrys** [reply]
@korycins What if we offered fully discounted line totals in addition to the ones that only contain line-level discounts? The app could then pick the ones that make more sense for the use case.

Alternatively, we could parametrize the subscription query to allow the subscriber to decide how and if discounts are distributed.

**Reply by korycins** [reply]
Yeah, that sounds great. I will create the task for this
